### PR TITLE
Push Publishing

### LINKED ISSUE #23: Push Publishing for trusted sources
Implement the mediachain/node/push protocol to allow trusted publishers to push statements to stable nodes.

Requires #12 to implement.

### PR DESCRIPTION
**Pull request by vyzo**
Implement the `/mediachain/node/push` protocol for accepting pushes from authorized peers.
Semantics are those of an inverted merge: the source pushes statements, and the receiver merges them and requests objects by reusing the merge protocol implementation.

Closes #23 

Implementation:
- [X] protobufs
- [X] receiver implementation
- [X] sender implementation
- [X] peer authorization/API for authorizing peers
- [X] namespace filtering in incoming stream

Also: changes mc/nat public ip detection service to ifconfig.co, as ifconfig.me appears to be busted.

### Examples
- Pushing to a remote peer which has authorized us
```
$ mcclient query "SELECT COUNT(*) FROM images.dpla"
1000
$ mcclient query -r QmSbgwkKxtrBFDoGGwoKqH7soYiL64xpMK5oEYdiRnQGJp "SELECT COUNT(*) FROM images.dpla"
0
$ curl -H "Content-Type: application/text" -d 'SELECT * FROM images.dpla' http://127.0.0.1:9002/push/QmSbgwkKxtrBFDoGGwoKqH7soYiL64xpMK5oEYdiRnQGJp
1000
1001
$ mcclient query -r QmSbgwkKxtrBFDoGGwoKqH7soYiL64xpMK5oEYdiRnQGJp "SELECT COUNT(*) FROM images.dpla"
1000
```

- Using the `/auth` api:
```
$ mcclient id
Peer ID: QmeBkfxcaBfA9pvzivRwhF2PM7sXpp4HHQbp7jfTkRCWEa
Publisher ID: 4XTTMADSKQUN3jkeZngbtuE35w9y5YnDTicVTeeji7N2Npkey
Info: vyzo testing

$ curl http://127.0.0.1:9002/auth
{}

# Authorize a peer (just self here)
$ curl -H "Content-Type: application/text" -d '*' http://127.0.0.1:9002/auth/QmeBkfxcaBfA9pvzivRwhF2PM7sXpp4HHQbp7jfTkRCWEa
OK
$ curl http://127.0.0.1:9002/auth
{"QmeBkfxcaBfA9pvzivRwhF2PM7sXpp4HHQbp7jfTkRCWEa":["*"]}
$ curl -H "Content-Type: application/text" -d '*,foo.*,bar.baz' http://127.0.0.1:9002/auth/QmeBkfxcaBfA9pvzivRwhF2PM7sXpp4HHQbp7jfTkRCWEa
OK
$ curl http://127.0.0.1:9002/auth
{"QmeBkfxcaBfA9pvzivRwhF2PM7sXpp4HHQbp7jfTkRCWEa":["*","foo.*","bar.baz"]}
$ curl http://127.0.0.1:9002/auth/QmeBkfxcaBfA9pvzivRwhF2PM7sXpp4HHQbp7jfTkRCWEa
*,foo.*,bar.baz

# revoke authorization
$ curl -H "Content-Type: application/text" -d '' http://127.0.0.1:9002/auth/QmeBkfxcaBfA9pvzivRwhF2PM7sXpp4HHQbp7jfTkRCWEa
OK
$ curl http://127.0.0.1:9002/auth
{}
```

### CODE REVIEW

File: mcnode/node.go [hunk 1]
```diff
@@ -323,3 +325,8 @@ func (node *Node) doShutdown() {
 	node.ds.Close()
 	os.Exit(0)
 }
+
+func (node *Node) authorizePush(pid p2p_peer.ID, ns []string) bool {
```
Comment (parkan) [inline mcnode/node.go]: let's just make a little table of permitted peers:namespaces for now?

**Comment by parkan** [comment]
Looking good so far

File: mcnode/node.go (hunk 1 above)
Comment (vyzo) [inline mcnode/node.go]: yeah, more like a table to a list of namespaces (with wildcard support)

File: mcnode/node.go [hunk 2]
```diff
@@ -323,3 +343,106 @@ func (node *Node) doShutdown() {
 	node.ds.Close()
 	os.Exit(0)
 }
+
+func (auth *PeerAuth) fromJSON(rmap map[string]interface{}) error {
+	auth.mx.Lock()
+	defer auth.mx.Unlock()
```
Comment (parkan) [inline mcnode/node.go]: this just executes when it goes out of scope? weird. I guess it's like a `finally` a bit?

File: mcnode/node.go [hunk 3]
```diff
@@ -323,3 +343,106 @@ func (node *Node) doShutdown() {
...
+}
+
+func (auth *PeerAuth) authorizeAllow(rules []string, ns string) bool {
+	for _, rule := range rules {
+		switch {
+		case rule == "*":
+			return true
+
+		case strings.HasSuffix(rule, ".*"):
+			if strings.HasPrefix(ns, rule[:len(rule)-2]) {
```
Comment (parkan) [inline mcnode/node.go]: what about `foo.bar.*.*`?

File: mc/nat.go [hunk 4]
```diff
@@ -119,7 +119,7 @@ func NATConfigFromString(str string) (cfg NATConfig, err error) {
 }
 
 func GetPublicIP() (string, error) {
-	res, err := http.Get("http://ifconfig.me/ip")
+	res, err := http.Get("http://ifconfig.co/ip")
```
Comment (vyzo) [inline mc/nat.go]: Sure, that's probably a good idea.
The code running ifconfig.co is available at github: https://github.com/martinp/ipd
So we can run it at `ifconfig.medichain.io` and have fallback to `ifconfig.co`.

I'll open an enhancement issue and add support once we have the ifconfing up.

File: mcnode/node.go (hunk 2 above)
Comment (vyzo) [inline mcnode/node.go]: It's more like `unwind-protect` at function-scope.
Each function has an unwind stack, where defers pushes thunks. And at the function epilogue, the unwind stack is poped and the thunks are executed.

File: mcnode/node.go (hunk 3 above)
Comment (vyzo) [inline mcnode/node.go]: Uhm, I don't think it's worth the trouble.
I kept it simple and used the same wildcard semantics we have in MCQL.
If it proves to be insufficient for some use case, we can generalize to support regexps.

**Pull request merged**